x/rollapp/keeper/block_height_to_finalization_queue.go

```diff
@@ -2,65 +2,114 @@ package keeper
+	// finalize pending states
+	for _, stateInfoIndex := range queue.FinalizationQueue {
+		if _, failed := failedRollapps[stateInfoIndex.RollappId]; failed {
+			// if the rollapp has already failed to finalize for at least one state index,
+			// skip all subsequent state changes for this rollapp
+			continue
 		}
-		k.RemoveBlockHeightToFinalizationQueue(ctx, blockHeightToFinalizationQueue.CreationHeight)
+		if err := k.finalizeStateForIndex(ctx, stateInfoIndex); err != nil {
+			// record the rollapp that had a failed state change at current height
+			failedRollapps[stateInfoIndex.RollappId] = stateInfoIndex.Index
+		}
+	}
+	k.updateQueueForHeight(ctx, queue, failedRollapps)
+}
+
+func (k Keeper) finalizeStateForIndex(ctx sdk.Context, stateInfoIndex types.StateInfoIndex) error {
+	// if this fails, no state change will happen
+	err := osmoutils.ApplyFuncIfNoError(ctx, func(ctx sdk.Context) error {
+		return k.finalizePending(ctx, stateInfoIndex)
+	})
+	if err != nil {
+		// TODO: think about (non)recoverable errors and how to handle them accordingly
+		k.Logger(ctx).Error(
+			"failed to finalize state",
+			"rollapp", stateInfoIndex.RollappId,
+			"index", stateInfoIndex.Index,
+			"error", err.Error(),
+		)
+		return fmt.Errorf("finalize state: %w", err)
 	}
 	return nil
 }
 
+func (k *Keeper) finalizePendingState(ctx sdk.Context, stateInfoIndex types.StateInfoIndex) error {
+	stateInfo := k.MustGetStateInfo(ctx, stateInfoIndex.RollappId, stateInfoIndex.Index)
+	if stateInfo.Status != common.Status_PENDING {
+		panic(fmt.Sprintf("invariant broken: stateInfo is not in pending state: rollapp: %s: status: %s", stateInfoIndex.RollappId, stateInfo.Status))
+	}
+	stateInfo.Finalize()
+	// update the status of the stateInfo
+	k.SetStateInfo(ctx, stateInfo)
+	// update the LatestStateInfoIndex of the rollapp
+	k.SetLatestFinalizedStateIndex(ctx, stateInfoIndex)
+	// call the after-update-state hook
+	keeperHooks := k.GetHooks()
+	err := keeperHooks.AfterStateFinalized(ctx, stateInfoIndex.RollappId, &stateInfo)
+	if err != nil {
+		return fmt.Errorf("after state finalized: %w", err)
+	}
+
+	// emit event
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(types.EventTypeStatusChange,
+			stateInfo.GetEvents()...,
+		),
+	)
+	return nil
+}
+
+func (k Keeper) updateQueueForHeight(ctx sdk.Context, blockHeightToFinalizationQueue types.BlockHeightToFinalizationQueue, failedRollapps map[string]uint64) {
+	// remove the blockHeightToFinalizationQueue if all the rollapps' states are finalized
+	if len(failedRollapps) == 0 {
+		k.RemoveBlockHeightToFinalizationQueue(ctx, blockHeightToFinalizationQueue.CreationHeight)
+		return
+	}
+	// remove from the queue only the rollapps that were successfully finalized at all indices.
+	// for failed rollapps at current height, all state indices including and following the failed index will stay in the queue
```

Comment:
```suggestion
	// remove from the queue the rollapps that were successfully finalized at all indices.
	// for failed rollapps at current height, all state indices including and following the failed index will stay in the queue
```

Reply:
Is this really a blocker?

Reply:
I mean it's wrong and makes it harder to understand tbh

Reply:
you can see it confused omri as well
https://github.com/dymensionxyz/dymension/pull/966#discussion_r1680629576